nvflare/app_common/streamers/container_streamer.py

```diff
@@ -0,0 +1,255 @@
+
+
+class _EntryConsumer(ObjectConsumer):
+    def __init__(self, stream_ctx: StreamContext):
+        self.logger = get_obj_logger(self)
+        container_type = stream_ctx.get(_CTX_TYPE)
+        container_class = load_class(container_type)
+        self.container = container_class()
+        self.size = stream_ctx.get(_CTX_SIZE)
+
+    def consume(
+        self,
+        shareable: Shareable,
+        stream_ctx: StreamContext,
+        fl_ctx: FLContext,
+    ) -> Tuple[bool, Shareable]:
+
+        entry = shareable.get(_KEY_ENTRY)
+        try:
+            if isinstance(self.container, dict):
+                key, value = entry
+                self.container[key] = value
+            elif isinstance(self.container, set):
+                self.container.add(entry)
+            else:
+                self.container.append(entry)
+        except Exception as ex:
+            error = f"Unable to add entry ({type(entry)} to container ({type(self.container)}"
+            self.logger.error(error)
+            raise ValueError(error)
+
+        last = shareable.get(_KEY_LAST)
+        if last:
+            # Check if all entries are added
+            if self.size != len(self.container):
+                err = f"Container size {len(self.container)} does not match expected size {self.size}"
+                self.logger.error(err)
+                raise ValueError(err)
+            else:
+                stream_ctx[_CTX_RESULT] = self.container
+                return False, make_reply(ReturnCode.OK)
+        else:
+            # continue streaming
+            return True, make_reply(ReturnCode.OK)
+
+    def finalize(self, stream_ctx: StreamContext, fl_ctx: FLContext):
+        self.logger.debug(f"Container streaming is done for container type {type(self.container)}")
+
+
+class _EntryConsumerFactory(ConsumerFactory):
+
+    def get_consumer(self, stream_ctx: StreamContext, fl_ctx: FLContext) -> ObjectConsumer:
+        return _EntryConsumer(stream_ctx)
+
+
+class _EntryProducer(ObjectProducer):
+    def __init__(self, container, entry_timeout):
+        self.logger = get_obj_logger(self)
+        if not container:
+            error = "Can't stream empty container"
+            self.logger.error(error)
+            raise ValueError(error)
+
+        self.container = container
+        if isinstance(container, dict):
+            self.iterator = iter(container.items())
+        else:
+            self.iterator = iter(container)
+        self.size = len(container)
+        self.count = 0
+        self.last = False
+        self.entry_timeout = entry_timeout
+
+    def produce(
+        self,
+        stream_ctx: StreamContext,
+        fl_ctx: FLContext,
+    ) -> Tuple[Shareable, float]:
+
+        try:
+            entry = next(self.iterator)
+            self.count += 1
+            self.last = self.count >= self.size
+        except StopIteration:
+            self.logger.error(f"Producer called too many times {self.count}/{self.size}")
```

Comment:
This condition can only happen when the container is empty or programming error.
You should just return (None, 0) to ensure aborting, instead of returning a Shareable.

Reply:
Changed.